astrodata/fits.py

```diff
@@ -1578,14 +1602,49 @@ def generate_boxes(shape, kernel):
     # The Astropy logger's "INFO" messages aren't warnings, so have to fudge
     log_level = astropy.logger.conf.log_level
     astropy.log.setLevel(astropy.logger.WARNING)
-    for coords in generate_boxes(shape, kernel):
-        # The coordinates come as ((x1, x2, ..., xn), (y1, y2, ..., yn), ...)
-        # Zipping them will get us a more desirable ((x1, y1, ...), (x2, y2, ...), ..., (xn, yn, ...))
-        # box = list(zip(*coords))
-        section = tuple([slice(start, end) for (start, end) in coords])
-        result.set_section(section, fn(element.window[section] for element in sequence))
-        gc.collect()
-    astropy.log.setLevel(log_level)  # and reset
+
+    boxes = generate_boxes(shape, kernel)
+
+    try:
+        for coords in boxes:
+            section = tuple([slice(start, end) for (start, end) in coords])
+            out = func([element.window[section] for element in sequence],
+                       **kwargs)
+            result.set_section(section, out)
+
+            # propagate additional attributes
+            if out.meta.get('other'):
+                for k, v in out.meta['other'].items():
+                    if len(boxes) > 1:
+                        result.meta['other'][k, coords] = v
+                    else:
+                        result.meta['other'][k] = v
+
+            gc.collect()
+    finally:
+        astropy.log.setLevel(log_level)  # and reset
+
+    # Now if the input arrays where splitted in chunks, we need to gather
+    # the data arrays for the additional attributes.
+    other = result.meta['other']
+    if other:
+        if len(boxes) > 1:
+            for (name, coords), obj in list(other.items()):
+                if not isinstance(obj, NDData):
+                    raise ValueError('only NDData objects are handled here')
+                if name not in other:
+                    other[name] = NDDataObject(np.empty(shape,
+                                                        dtype=obj.data.dtype))
+                section = tuple([slice(start, end) for (start, end) in coords])
+                other[name].set_section(section, obj)
+                del other[name, coords]
+
+        for name in other:
+            # To set the name of our object we need to save it as an ndarray,
+            # otherwise for a NDData one AstroData would use the name of the
+            # AstroData object.
+            other[name] = other[name].data
+            result.meta['other_header'][name] = fits.Header({'EXTNAME': name})
```

Comment:
So the most tricky part is here, because `windowedOp` can split the NDData objects in blocks. So in the loop I save the additional NDData objects (in `.meta['other']`) with their coordinates, and then I recreate the final object. But this way has some limitations, like I cannot guess the shape so I use the input one, but because of that I cannot save the full cube of the individual masks. I guess saving the "rejection map", i.e. the sum over the num_img axis, is good enough ?

Also another issue here: if I save a NDData object, AstroData has no way to set its extension name... But if I save a ndarray I can set its EXTNAME with the header:
https://github.com/GeminiDRSoftware/DRAGONS/blob/c952cd99d5a39a384b35d3ae9b64812bdf80bd2b/astrodata/fits.py#L1047-L1053